src/solvers/strings/array_pool.cpp

```diff
@@ -115,6 +115,54 @@ array_string_exprt array_poolt::make_char_array_for_char_pointer(
   return to_array_string_expr(insert_result.first->second);
 }
 
+/// Given an array_string_exprt, get the size of the underlying array. If that
+/// size is undefined, create a new symbol for the size.
+/// Then add an entry from `array_expr` to that size in the `length_of_array`
+/// map.
+///
+/// If `array_expr` is an `if_exprt`, recursively add entries for the parts of
+/// the `if_exprt`, but not for the `if_exprt`s themselves.
+///
+/// This function should only be used at the creation of the
+/// `array_string_exprt`s, as it is the only place where we can reliably refer
+/// to the size in the type of the array.
+static void attempt_assign_length_from_type(
+  const array_string_exprt &array_expr,
+  std::unordered_map<array_string_exprt, exprt, irep_hash> &length_of_array,
+  symbol_generatort &symbol_generator)
+{
+  if(const auto &if_expr = expr_try_dynamic_cast<if_exprt>((exprt)array_expr))
```

Comment:
Why the explicit cast (and to a non-reference-type?)

Reply:
You mean the `(exprt)`? Curiously `exprt_try_dynamic_cast` doesn't work directly on `array_string_exprt`.

Reply:
Oh wait, this is obvious -- an `array_string_exprt` is certainly not an `if_exprt`, this can't ever be run.

Reply:
There's one other case with the same cast and the same unreachable code. This suggests writing tests to exercise this code!

Reply:
I don't think it's unreachable. `array_string_exprt`s can have `ID_if` AFAIK. However you make a good point with the test coverage, as we don't seem to have anything to cover these lines so for now I might just put an assertion.

Reply:
So in this place i replaced the if with an assertion because I can't find any test to cover this part. However in the other place where I use the condition, I left it in because it's often the case that the `array_string_exprt` is an if. Therefore that code is well covered by regression tests.

Reply:
Aha, my mistake, I thought `array_string_exprt` constrained the expression to have a particular id, and therefore not be an `if_exprt`, but actually it just constrains the type of a general expression. The implementation of the expr-casting stuff, though, assumes this sort of trait class doesn't exist, and so casts are only possible to children of the current type. That ought to be fixed, but is obviously out of scope for this PR, so approving.